### DISCUSSION TITLE
Information and how-to for the harbor-http (get-metadata)

### DISCUSSION THREAD

**Discussion by carrzkiss** [savonet/liquidsoap#3857] [General]
Thanks to, [Vito](https://github.com/vitoyucepi) for the information on [harbor-http](https://www.liquidsoap.info/doc-2.2.4/harbor_http.html).
Looking at the information and examples, I really like the [get-metadata](https://www.liquidsoap.info/doc-2.2.4/harbor_http.html#get-metadata) section for right now, but I have no idea how to implement and test it.
Do I place the code in my 101.liq file, or is it placed somewhere else, and a call made to the file with the link address to the media file?

Any information on this would be great.
Thank you.
Wayne

**Comment by gAlleb** [comment]
Hi @carrzkiss!

In order for `get-metadata` to work I could suggest you to do the following:

1. Insert the following code to you `101.liq` config just after `music = playlist(mode = 'normal', reload = 1, reload_mode = 'seconds', "https://www.example.com/DJ/101.asp")`:

```ruby
meta = ref([])
music.on_metadata(fun (m) -> meta := m)
def get_meta(_, response) =
  response.json(meta())
end
harbor.http.register(port=8007, method="GET", "/metadata", get_meta)
```

Then, when you will navigate to `http://localhost:8007/metadata` - you will see all the metadata available for your current track.

If you want you can proxy `http://localhost:8007/metadata` in you nginx configuration file to get a nicer url. E.g.:

Paste this inside `/etc/nginx/sites-enabled/YOUR.conf` within a `server` block (server {}) and restart nginx (`sudo systemctl reload nginx`):
```
location /metadata {
            proxy_pass  http://127.0.0.1:8007/metadata;
}
```

Now, when you navigate to `http://your-site-adress.com/metadata` - you will see your metadata.

**Reply by gAlleb** [reply]
Also this way you will get a tone of metadata you don't need. If you want only the basic one replace

this:

```ruby
meta = ref([])
music.on_metadata(fun (m) -> meta := m)
def get_meta(_, response) =
  response.json(meta())
end
harbor.http.register(port=8007, method="GET", "/metadata", get_meta)
```

with this (courtesy of @vitoyucepi):

```ruby
def get_meta(_)

  def null_float(f)
    f == infinity ? null() : f
  end

  def null_list(key, _list)
    list.assoc.mem(key, _list) ? list.assoc(key, _list) : null()
  end


  m = music.last_metadata() ?? []

  data = {
    now_playing = {
      on_air_timestamp = null_list("on_air_timestamp", m),
      on_air = null_list("on_air", m),
      duration = null_float(source.duration(music)),
      elapsed = null_float(source.elapsed(music)),
      remaining = null_float(source.remaining(music)),
      filename = null_list("filename", m),
      artist = null_list("artist", m),
      title = null_list("title", m),
      album = null_list("album", m),
      genre = null_list("genre", m)
     }
  }

 try
  http.response(
    status_code=200,
    headers=[
                ("Content-Type", "application/json")
            ],
    content_type="application/json; charset=UTF-8",
    data=json.stringify(data, compact=true) ^ "\n"
  )

 catch error do
  log.severe(error.message, label="http")
  http.response(
    status_code=500,
    content_type="application/json; charset=UTF-8",
    data='{"status":"error","message":"Failed to serialize response data"}' ^ "\n"
  )
  end
end

harbor.http.register.simple("/metadata", get_meta, port=8007, method="GET")
```

Now you'll get only artist,title,album,timestamp... so on.

**Reply by carrzkiss** [reply]
Thank you, so much, Stefan.
So all this will go in the 101.liq file?

I will test this once I am back in.
If this works with my current setup it is going to be a game changer.
If does not work then I will implement the local directory.

**Reply by carrzkiss** [reply]
OK. added the code to the liq file.
No errors, which is good.

How to access the page?
I've opened port 8007 TCP in PFSense.
I've tried
`http://stream.example.com/metadata`
`http://stream.example.com:8007/metadata`
`http://stream.example.com:8007/hls/metadata`
`http://stream.example.com:8007/hls/101/metadata`
Neither are working.

**Reply by carrzkiss** [reply]
I can get to it locally.
And it shows some info. I might have to do some tweaking.
`http://192.168.5.10:8007/metadata`
I just changed it so I can access each individual file.
`http://192.168.5.10:8007/101/metadata`
`harbor.http.register.simple("/101/metadata", get_meta, port=8007, method="GET")`

**Reply by carrzkiss** [reply]
I noticed that the filename was the encoded partial filename.
I ran this.
```
meta = ref([])
music.on_metadata(fun (m) -> meta := m)
def get_meta(_, response) =
  response.json(meta())
end
harbor.http.register(port=8007, method="GET", "/metadata", get_meta)
```
And found. `initial_uri`
So, I changed the 
`filename = null_list("filename", m),`
to
`filename = null_list("initial_uri", m),`
And this gives me the file URL.
`https://demo.example.com/host/Files/Short/05292023072448785.mp3`
This is the main thing I needed. If I had the actual filename, I can manipulate anything on the server side. This is wonderful.

I am still trying to figure out why I cannot access with the domain URL.
I added a new section in the `sites-enabled` specifically for port 8007.

```
server {
listen 8007;
location /metadata {
            proxy_pass  http://127.0.0.1:8007/metadata;
}
}
```
But it still does not allow the domain URL to get through.

**Reply by vitoyucepi** [reply]
> This is the main thing I needed. If I had the actual filename, I can manipulate anything on the server side. This is wonderful.

The `filename` and `initial_uri` are not very reliable sources, so be warned.

> ```nginx
> location /metadata {
>   proxy_pass http://127.0.0.1:8007/metadata;
> }
> ```

Same problem as here https://github.com/savonet/liquidsoap/discussions/3844#discussioncomment-8968356
```nginx
location /metadata {
  proxy_pass http://127.0.0.1:8007/;
}
```

> ```nginx
> listen 8007;
> location /metadata {
>   proxy_pass http://127.0.0.1:8007/;
> }
> ```

This will not work, you cannot bind on the same port.
Also consider caching the response for at least a second.
You will also want to change `settings.harbor.bind_addrs := ["127.0.0.1"]` to avoid exposing the liquidsoap port.

**Reply by carrzkiss** [reply]
Then that would explain the error I got.
OK.

**Reply by carrzkiss** [reply]
While this go in the liq file?
`settings.harbor.bind_addrs := ["127.0.0.1"]`

**Reply by carrzkiss** [reply]
So far, in the testing I have done, all information is accurate for `initial_uri`.
I can understand where the `filename` would not be as accurate as it keeps changing as there are multiple temp files that create the single stream, but the `initial_uri` is just one name and never changes, just like the artist, title, album, etc.

**Reply by gAlleb** [reply]
> I can understand where the filename would not be as accurate as it keeps changing as there are multiple temp files that create the single stream

That's not accurate because `filename` is not connected to the segments generated by HLS output. It's connected to the metadata. Same goes for `initial_uri`. Say, you have playlist with song in `/home/radio/my_song.mp3`. Then you'll get
`filename` - `/home/radio/my_song.mp3`
`initial_uri` - `/home/radio/my_song.mp3`

`initial_uri` can change depending of what else you do. E.g. you can `annotate` your files with different protocols or custom metadata, then you will end up getting `initial_uri` looking like this: `annotate:custom_="custom":protocol:/home/radio/my_song.mp3`

**Reply by vitoyucepi** [reply]
@gAlleb,
And they both have the same fatal flaw. By the way, what would the filename be if the file was remote?

**Reply by gAlleb** [reply]
Something like `/tmp/liq-process1a7527.mp3` :)

**Reply by carrzkiss** [reply]
So far, the `initial_uri` (Named it `filename` in the script) has not given me any errors.
`https://demo.example.com/host/Files/Lizzy_Borden/1989-Master_Of_Disguise/02082024183450436.mp3`
The `02082024183450436.mp3` will be the same as the title, artist, and all the other data in the file.
If I see an issue, I will address it when it occurs and find another means of obtaining the needed information.
However, with over a thousand tests (Songs played), the `initial_uri` has been accurate every time.

**Comment by carrzkiss** [comment]
Is there a log file entry for every time the `/metadata` is hit? (Viewed)
